extensions/extensionservice/org.eclipse.smarthome.extensionservice.marketplace/src/main/java/org/eclipse/smarthome/extensionservice/marketplace/internal/MarketplaceExtensionService.java

```diff
@@ -0,0 +1,236 @@
+        List<Extension> exts = new ArrayList<>(nodes.size());
+        for (Node node : nodes) {
+            if (node.id == null) {
+                // workaround for bug https://bugs.eclipse.org/bugs/show_bug.cgi?id=512493
+                continue;
+            }
+            Extension ext = convertToExtension(node);
+            if (ext != null) {
+                exts.add(ext);
+            }
+        }
+        return exts;
+    }
+
+    private MarketplaceExtension convertToExtension(Node node) {
+        String extId = getExtensionId(node);
+        if ("binding".equals(node.packagetypes)) {
+            MarketplaceExtension ext = new MarketplaceExtension(extId, "binding", node.name, node.version,
+                    node.supporturl, isInstalled(extId), node.shortdescription, null, node.image, node.updateurl);
+            return ext;
+        } else if ("rule_template".equals(node.packagetypes)) {
+            String version = StringUtils.isNotEmpty(node.version) ? node.version : "1.0";
+            MarketplaceExtension ext = new MarketplaceExtension(extId, "ruletemplate", node.name, version,
+                    node.supporturl, isInstalled(extId), node.shortdescription, null, node.image, node.updateurl);
+            return ext;
+        } else {
+            return null;
+        }
+    }
+
+    @Override
+    public Extension getExtension(String id, Locale locale) {
+        for (Extension extension : getExtensions(locale)) {
+            if (extension.getId().equals(id)) {
+                return extension;
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public List<ExtensionType> getTypes(Locale locale) {
+        return TYPES;
+    }
+
+    @Override
+    public void install(String extensionId) {
+        Extension ext = getExtension(extensionId, null);
+        if (ext instanceof MarketplaceExtension && !ext.isInstalled()) {
+            MarketplaceExtension mpExt = (MarketplaceExtension) ext;
+            String type = getType(extensionId);
+            switch (type) {
+                case "ruletemplate":
+                    String url = mpExt.getDownloadUrl();
+                    try {
+                        String template = getTemplate(url);
+                        marketplaceRuleTemplateProvider.addTemplateAsJSON(extensionId, template);
+                    } catch (ParsingException e) {
+                        postFailureEvent(extensionId, "Template is not valid.");
+                        logger.error("Rule template from marketplace is invalid: {}", e.getMessage());
+                    } catch (IOException e) {
+                        postFailureEvent(extensionId, "Template cannot be downloaded.");
+                        logger.error("Rule template from marketplace cannot be downloaded: {}", e.getMessage());
+                    }
+                    break;
+                default:
+                    postFailureEvent(extensionId, "Id not known.");
```

Comment:
In `uninstall` you return, so no post(Un)installedEvent is called. Do you want to call `return` here, too?